packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart

```diff
@@ -224,23 +239,60 @@ abstract class RenderSliverMultiBoxAdaptor extends RenderSliver
 
   bool _debugAssertChildListLocked() => childManager.debugAssertChildListLocked();
 
-  @override
-  void insert(RenderBox child, { RenderBox after }) {
-    assert(!_keepAliveBucket.containsValue(child));
-    super.insert(child, after: after);
-    assert(firstChild != null);
-    assert(() {
+  /// Verify the child list index is in strictly increased order.
+  void _debugVerifyChildOrder(){
+    if (_integrityEnabled) {
       int index = indexOf(firstChild);
       RenderBox child = childAfter(firstChild);
       while (child != null) {
         assert(indexOf(child) > index);
         index = indexOf(child);
         child = childAfter(child);
       }
+    }
+  }
+
+  @override
+  void insert(RenderBox child, { RenderBox after }) {
+    assert(!_keepAliveBucket.containsValue(child));
+    super.insert(child, after: after);
+    assert(firstChild != null);
+    assert(() {
+      _debugVerifyChildOrder();
       return true;
     }());
   }
 
+  @override
+  void move(RenderBox child, { RenderBox after }) {
+    // There are two scenarios:
+    //
+    // 1, The child is in childList maintained by ContainerRenderObjectMixin.
+    // We can call the super.move and update parentData with new slot.
+    //
+    // 2, The child is in _keepAliveBucket.
+    // In this case, the child is no longer in the childList.
+    // We need to take it out from bucket, readopt it, and insert back to childList.
+    final SliverMultiBoxAdaptorParentData childParentData = child.parentData;
+    if (!_keepAliveBucket.containsKey(childParentData.index)) {
+      super.move(child, after: after);
+      // The parent class does not update the slot.
+      // We have to explictly update it here.
+      childManager.didAdoptChild(child);
+    } else {
+      _keepAliveBucket.remove(childParentData.index);
```

Comment:
we don't know if the new index is visible in viewport or not. It is performlayout that determines that.
if it is in viewport, it will not be stored back to _keepAliveBucket and remain in childlist.
if not, it will be stored back to _keepAliveBucket.
either case, I think we should rely on performlayout to do the right thing

Reply:
Let's talk about this one in person when you have a second.